docs/download-upload-clone/storage/high-level-design.md

```diff
@@ -0,0 +1,163 @@
+for temporary storage of shapefiles uploaded by users.
+
+Staging/production/demo instances, on the other hand, will normally use cloud
+blob storage (Azure Blob Storage) for performance reasons.
+
+Whether local or remote/cloud storage is used, the backend code will use a
+single storage interface, configured to work with suitable implementations
+through instance config.
+
+## Production - cloud storage
+
+Azure Blob Storage is used to stream and store individual archive pieces.
+
+While archives are being created, pieces are stored privately, i.e. only the
+service account configured to allow the MarxanCloud backend to interact with
+Azure storage will have access to the relevant blobs.
+
+Each MarxanCloud instance makes use of a unique Azure storage account and
+related service account.
+
+Isolating tenants at the organization level may be desirable at a later stage
+for security, compliance (e.g. choice of Azure region) and billing reasons.
+
+When users request to actually download a complete archive, signed URLs are
+generated using Azure Shared Access Signature (SAS). These are nevertheless
+provided via redirection from local (API) URLs so that the eventual storage is
+transparent to the end user.
+
+## Development - local storage
+
+Local block volumes are used, transparently, as backing storage in development
+environments.
+
+More stringent size limits will apply, therefore project archive piece resolvers
+must be aware of size constraint and halt their work with a relevant error if
+available storage may not be sufficient to store a piece being prepared. This is
+so that filesystem saturation and related error states may be avoided.
+
+## Storage layout and metadata
+
+<bucketUrl>/<organizationId>/<projectId>/<exportId>
+
+The proposed URL schema is mainly meant to logically separate exports by parent
+project and organization.
+
+Blob lifecycle (for example, deleting exports at their expiration timestamp, or
+when their parent project is deleted from a MarxanCloud instance) should be
+managed through Azure blob tags.
+
+The files corresponding to pieces and final archive for each export should be
+tagged with:
+
+- `organizationId`
+- `projectId`
+- `scenarioId` (for scenario pieces)
+- `exportId`
+- `exportTimestamp` (timestamp indicating when the export was initiated;
+  identical for all pieces)
+- `expiryTimestamp` (used to programmatically remove expired exports and their
+  pieces)
+
+## Storage handling - flow
+
+This section expands with design details the summary list provided in the brief
+document.
+
+
+- storing one or more archive pieces as they are prepared and finalized
+
+For anything else than small metadata files, these pieces should be streamed to
+their eventual storage destination.
+
+- preparing a final archive file that includes all the finalized pieces
+
+This would typically be a zip (ZIP64) file. It may also be advisable to keep all
```

Comment:
I would ditch making the individual pieces available, at least for now. This is not only more work for us, but also potentially confusing for a user - what will you do with a single scenario piece, for example? I think this also slightly contradicts, in spirit, the checksum part, were we go out of our way to ensure the pieces are not manually modified.

Reply:
so, I whereas I would slightly lean on giving users a single zip file, I still think that keeping the single pieces available on blob storage until their expiration (or deletion if we support explicitly deleting an export) would be a good idea: if each individual piece is streamed to blob storage while being prepared, then - unless the user wants to download the archive locally, in which case we can prepare a zip for them - we could leave the pieces there, and if the user wants to duplicate a project, we wouldn't have to deal with any archive ourselves, just stream in the pieces, use to populate tables/etc, and this could be done piece by piece, with dependencies honoured where some piece needs to get in before others can.

why do you think this would be more work for us?